### ISSUE TITLE
Generic type that conditionally returns `never` can result in a `never` acting like `any` when used

### ISSUE THREAD

**Issue by itaylor** [microsoft/TypeScript#50090]
# Bug Report

### 🔎 Search Terms

generics, as keyword, type assertion, never

### 🕗 Version & Regression Information

I see the same behavior in versions 3.3 through 4.8.0 beta in the TS Playground

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about __the never type being used after being returned_______

### ⏯ Playground Link

[Playground link](https://www.typescriptlang.org/play?ts=4.8.0-beta&ssl=20&ssc=1&pln=1&pc=1#code/C4TwDgpgBAyg9gWwsAlkgzgOQgNwgJwB4AVAPigF4oAoKKYqCAD2AgDsATdKdYfFNgHMoAfh58Bg2nSgAueoxbsuUNgFcEAIwKjVuAtLry2+-AG5q1AGZq2AY1Rw2PRMjQQspgGJsSpABTorsAAFpLyxACU8vBIqBjYeERkUADe0vjIavjOQXFhQlAAhtyxbgmmfhYAvpZ2TrxQcADWlC5x7p5JPv4A5HnIBYK9kRYccGWhkv4to1AA9POKkA4QHIz4+HD48gCC+IIa7MBNVlCgkFD9EkK9UCjcbHAnJegogmxFmgA20MBwUDARXwRTiOjgZwu0F6ABEiqxegA6fwAJgAzAAWACskTqDROAHdgGcqAN4h5EgQegBGAAMo2o40mQ38RKscwWS2IYW46BCcDU33WBC2+AANFBNGoTuMPGxesBEXQ2fduFCoAADExJDUSqWE6B2IrOIHobgoE7-YpQGz2RzOULw5YQBzcIpQOGsESWW0OFBOKBM4Is9DyT0QSJpaToRGCZDh-wM2rUIA)

### 💻 Code

```ts
type SometimesNever<T> = 
  T extends string ? string
    : T extends number ? never
    : never;

function sometimesNeverFn<T>(something: T): SometimesNever<T> {
  return something as SometimesNever<T>;
}

const ok = sometimesNeverFn('something');
doSomething(ok); // expected/actual error: Argument of type 'string' is not assignable to parameter of type 'Date'.(2345)

const wtf = sometimesNeverFn(10);
doSomething(wtf);  // expected: This should error, but doesn't.  `wtf` is type `never` (visible in playground) but we can pass it to a function that expects a Date.

function doSomething(s: Date) {
  s.getDate();
}
```

### 🙁 Actual behavior
When a generic type that might be `never` is returned from a function by using a type assertion in the return, and the type resolves as `never`, the type is still usable to be passed as an argument to functions, almost as if it was an `any`.

### 🙂 Expected behavior

If the compiler is able to resolve a type as `never`, then the variable with that type should cause an error when passed as a function argument.

I had this occur on a vastly more complex set of generic types and functions that could sometimes be `never` (and correctly so, as the function did runtime checks and would `throw` if passed certain types), and it was very unintuitive that the type could be resolved as `never` but still returned and subsequently passed to another function without an error.

**Comment by MartinJohns** [comment]
That's working as intended. Nothing is assignable to `never`, but `never` is assignable to anything.

**Comment by itaylor** [comment]
Yeah, can see that even this works.
```ts
function alwaysNever(something: string | number): never {
  throw Error('err');
}

const ok = alwaysNever('something');
doSomething(ok);

const wtf = alwaysNever(10);
doSomething(wtf);  

function doSomething(s: Date) {
  s.getDate();
}
```

**Comment by itaylor** [comment]
Thanks for the explanation @MartinJohns.  I'll close the ticket.  It is very unintuitive, but I think I understand why it's happening.

**Comment by itaylor** [comment]
I think what I'm really wishing for here is something like what is proposed in #23689, a type that is neither assignable to anything nor assignable from anything.

**Comment by fatcerberus** [comment]
Yep, #23689 (`invalid` type) is most likely what you're really looking for.

> It is very unintuitive, but I think I understand why it's happening.

Yeah, so `never` is the bottom type which represents the "result" of infinite loops or exceptions, so you (and the compiler) are free to pretend it's whatever type you want.  This is the type-theoretically correct behavior, see https://github.com/microsoft/TypeScript/issues/49862#issuecomment-1180954393

tl;dr: If `doSomething(neverVal)` is reachable, your program is already ill-typed by definition.
